### PR TITLE
Support journal-upload HTTPS without key and certificate

### LINKED ISSUE #14886: 90 seconds delay during boot, trying to rename 2 interfaces to eno1.
I get a 90 seconds delay during boot. It seems that hanging trying to rename an interface. dmesg shows:
```
[    2.524990] igb 0000:01:00.0: added PHC on eth0
[    2.524996] igb 0000:01:00.0: Intel(R) Gigabit Ethernet Network Connection
[    2.525002] igb 0000:01:00.0: eth0: (PCIe:2.5Gb/s:Width x1) XX:XX:XX:XX:XX:43
[    2.525067] igb 0000:01:00.0: eth0: PBA No: 000300-000
[    2.525072] igb 0000:01:00.0: Using MSI-X interrupts. 4 rx queue(s), 4 tx queue(s)
[...]
[    2.554624] igb 0000:02:00.0: added PHC on eth1
[    2.554630] igb 0000:02:00.0: Intel(R) Gigabit Ethernet Network Connection
[    2.554636] igb 0000:02:00.0: eth1: (PCIe:2.5Gb/s:Width x1) XX:XX:XX:XX:XX:44
[    2.554702] igb 0000:02:00.0: eth1: PBA No: 000300-000
[    2.554707] igb 0000:02:00.0: Using MSI-X interrupts. 4 rx queue(s), 4 tx queue(s)
[    2.566821] igb 0000:01:00.0 eno1: renamed from eth0
[...]
[    2.606909] igb 0000:02:00.0 rename3: renamed from eth1
[...]
[    3.249856] sd 1:0:0:0: [sda] Attached SCSI disk
[   63.258688] systemd-udevd[101]: eth1: Worker [128] processing SEQNUM=1438 is taking a long time
[   93.193528] systemd-udevd[128]: eth1: Failed to rename network interface 3 from 'eth1' to 'eno1': File exists
[   93.205214] systemd-udevd[127]: Using default interface naming scheme 'v240'.
[...]
[  110.007404] IPv6: ADDRCONF(NETDEV_UP): eno1: link is not ready
[  111.113963] NFSD: starting 90-second grace period (net f0000098)
[  112.994184] igb 0000:01:00.0 eno1: igb: eno1 NIC Link is Up 1000 Mbps Full Duplex, Flow Control: RX/TX
[  112.994513] IPv6: ADDRCONF(NETDEV_CHANGE): eno1: link becomes ready
```

For some reason it's trying to rename both eth0 and eth1 to eno1. It seems to try to do:
eth0 -> eno1
eth1 -> rename3
eth1 -> eno1

I don't seem to have a file that gives the devices a persistent name.

ip addr shows:
```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eno1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether XX:XX:XX:XX:XX:43 brd ff:ff:ff:ff:ff:ff
    inet 192.168.0.7/24 brd 192.168.0.255 scope global eno1
       valid_lft forever preferred_lft forever
    inet6 XXXX/64 scope global dynamic mngtmpaddr 
       valid_lft 239811sec preferred_lft 67011sec
    inet6 fe80::XXXX/64 scope link 
       valid_lft forever preferred_lft forever
3: rename3: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default qlen 1000
    link/ether XX:XX:XX:XX:XX:44 brd ff:ff:ff:ff:ff:ff
```

I'm using Debian buster, version 241-7~deb10u3

### LINKED ISSUE #14919: Support journal-upload HTTPS without key and certificate
**Is your feature request related to a problem? Please describe.**

We like to upload journald-logs to SaaS HTTPS receiver without the need to create client certificates. 
All https uploads fail with this setting: 

```
[Upload]
URL=https://logsene-journald-receiver.sematext.com:443/LOGS_TOKEN
#ServerKeyFile=/etc/ssl/private/journal-upload.pem
#ServerCertificateFile=/etc/ssl/certs/journal-upload.pem
TrustedCertificateFile=/etc/ssl/certs/ca-certificates.crt
```

Error: 
```
failed: error reading X.509 key or certificate file
```

Setting no `ServerCertificateFile`  and `ServerKeyFile` file, the default value is used, but those files don't exist. 
https://github.com/systemd/systemd/blob/5ee69e144e57edd328abf85ef8790ed52217b930/src/journal-remote/journal-upload.c#L36-L37

journal-upload loads keyfile and certificate always for `https` URLs
See: https://github.com/systemd/systemd/blob/5ee69e144e57edd328abf85ef8790ed52217b930/src/journal-remote/journal-upload.c#L243-L250

**Describe the solution you'd like**

Change the condidtion in line 243: https://github.com/systemd/systemd/blob/5ee69e144e57edd328abf85ef8790ed52217b930/src/journal-remote/journal-upload.c#L243

```
if (arg_key || startswith(u->url, "https://")) {
...
```
to
```
if (arg_key) { 
...

// or when arg_key AND https URL is used

if (arg_key && startswith(u->url, "https://")) {
...

```

So keys are only loaded if the user has specified the keys - and not for every HTTPS URL. 

**Describe alternatives you've considered**
An alternative would be a setting to allow simple https uploads. 
In our case, we use a generated token in the URL.

### PR DESCRIPTION
**Pull request by hakman**
Most log management service are starting to allow direct upload from `systemd-journal-upload`. It is available in pretty much any recent Linux distro and it's very easy to set up.

To make it easier for the users to ship their logs to such services, `systemd-journal-upload` should support HTTPS connections without the need for key and certificate.

Fixes #14919.

### CODE REVIEW

**Comment by poettering** [comment]
That's quite some drastic change of defaults, no?

Why not introduce this in a less intrusive fashion, for example by understanding "-" as argument to these options as a request to not read these files?

**Comment by hakman** [comment]
In retrospect, I guess changing the defaults was not such a good idea. Though it was a good conversation starter :).
Thanks for taking the time to check and consider this change @poettering.

I spent some time thinking what could be a good value considering that `all` is used to disable cert checks.

Next I noticed that `config_parse_many_nulstr` is used to check if the args are actually valid files. Anything but a valid file name results in a NULL value. Pretty sure `all` becomes `NULL` also:
```
`/etc/systemd/journal-upload.conf:3: TrustedCertificateFile= path is not absolute, ignoring: all
```
https://github.com/systemd/systemd/blob/5ee69e144e57edd328abf85ef8790ed52217b930/src/journal-remote/journal-upload.c#L526

I would be happy to update the PR to accept `-` for the file names. Just a few questions:
1. should `-` also become an alternative to disable cert checks as it is `all`?
2. would it be a good idea to change the trust file default to `/etc/ssl/certs/ca-certificates.crt`?

**Comment by hakman** [comment]
@poettering I think the new changes are closer to what you are suggesting. Could you take another look? Thanks!

File: src/journal-remote/journal-upload.c [hunk 1]
```diff
@@ -240,14 +241,14 @@ int start_upload(Uploader *u,
...
-                if (arg_key || startswith(u->url, "https://")) {
+                if (!streq_ptr(arg_key, "-") && (arg_key || startswith(u->url, "https://"))) {
                         easy_setopt(curl, CURLOPT_SSLKEY, arg_key ?: PRIV_KEY_FILE,
                                     LOG_ERR, return -EXFULL);
                         easy_setopt(curl, CURLOPT_SSLCERT, arg_cert ?: CERT_FILE,
                                     LOG_ERR, return -EXFULL);
                 }
 
-                if (streq_ptr(arg_trust, "all"))
+                if (streq_ptr(arg_trust, "-") || streq_ptr(arg_trust, "all"))
```
Comment (keszybz) [inline src/journal-remote/journal-upload.c]: `STRPTR_IN_SET()`

File: src/journal-remote/journal-upload.c [hunk 2]
```diff
@@ -35,7 +36,7 @@
 
 #define PRIV_KEY_FILE CERTIFICATE_ROOT "/private/journal-upload.pem"
 #define CERT_FILE     CERTIFICATE_ROOT "/certs/journal-upload.pem"
-#define TRUST_FILE    CERTIFICATE_ROOT "/ca/trusted.pem"
+#define TRUST_FILE    CERTIFICATE_ROOT "/certs/ca-certificates.crt"
```
Comment (keszybz) [inline src/journal-remote/journal-upload.c]: Please make that a separate commit, with rationale. (I honestly don't know whether we should make this change. The reasoning needs to be recorded in the commit message.)

File: src/journal-remote/journal-upload.c [hunk 3]
```diff
@@ -515,12 +516,52 @@ static int perform_upload(Uploader *u) {
...
+        assert(data);
+
+        if (isempty(rvalue))
+                goto finalize;
+
+        n = strdup(rvalue);
+        if (!n)
+                return log_oom();
+
+        if (streq_ptr(n, "-"))
```
Comment (keszybz) [inline src/journal-remote/journal-upload.c]: `n` cannot be NULL here, so `streq` is enough.

File: src/journal-remote/journal-upload.c (hunk 2 above)
Comment (hakman) [reply src/journal-remote/journal-upload.c]: Agreed.

File: src/journal-remote/journal-upload.c (hunk 3 above)
Comment (hakman) [reply src/journal-remote/journal-upload.c]: Nice find. Updated.

**Comment by hakman** [comment]
Thanks @keszybz. I updated the PR with your suggestions.

**Comment by otisg** [comment]
@hakman Looks like one check failed.

File: man/systemd-journal-upload.service.xml [hunk 4]
```diff
@@ -184,7 +188,7 @@
 
         <listitem><para>
           Takes a path to a SSL CA certificate file in PEM format,
-          or <option>all</option>. If <option>all</option> is set,
+          or <option>-/all</option>. If <option>-/all</option> is set,
```
Comment (keszybz) [inline man/systemd-journal-upload.service.xml]: `<option>-</option>/<option>all</option>`

**Comment by hakman** [comment]
`autopkgtest` tests seem to timeout randomly. I don't think they are related to this PR. Any tips?

**Comment by keszybz** [comment]
Yeah, tests time out occasionally. Just ignore any failures which seem to be unrelated.

**Comment by keszybz** [comment]
Maybe fix the formatting as suggested above.

**Comment by hakman** [comment]
Thanks @keszybz. Formatting is also fixed.
